File: Kotlin/ByTier/F/2123_F_Minimize_Fixed_Points.kt
// Problem: CF 2123 F - Minimize Fixed Points
// https://codeforces.com/contest/2123/problem/F

/*
F. Minimize Fixed Points

Algorithms/Techniques:
- Sieve of Eratosthenes for precomputing smallest prime factor (SPF)
- Grouping elements by their smallest prime factor to form cycles
- Greedy assignment of elements in groups to minimize fixed points

Time Complexity: O(N log N) where N is the maximum value of n (10^5).
Space Complexity: O(N) for storing SPF array and other auxiliary structures.

This code finds a "good" permutation of length n such that:
- For each position i (2 <= i <= n), gcd(p[i], i) > 1
- The number of fixed points is minimized across all good permutations.

Approach:
1. Precompute the smallest prime factor (SPF) for numbers up to N using sieve.
2. For each input n, group numbers from 1 to n based on their SPF.
3. Within each group, assign elements in a cyclic manner to avoid fixed points.
4. This ensures that no element maps to itself and gcd(p[i], i) > 1 for all i >= 2.
*/

import kotlin.math.*
fun IntArray.mp(_x : Int) : Int {
    var ans = 1
    var x = _x
    while (x > 1) {
        ans = max(ans, this[x]) // Track the maximum prime factor encountered
        x /= this[x] // Move to next factor by dividing with SPF
    }
    return ans
}
fun main() {
    val N = 100001
    val lp = IntArray(N) // Array storing smallest prime factor for each number
    val p = mutableListOf<Int>() // List of primes found so far
    for (i in 2..N - 1) {
        if (lp[i] == 0) { // i is a prime
            lp[i] = i
            p.add(i)
        }
        // Mark multiples of current prime with this prime as their smallest factor
        p.takeWhile {it <= lp[i] && it * i < N}.forEach {lp[it * i] = it}
    }
    repeat(readln().toInt()) {
        val n = readln().toInt()
        val a = IntArray(n) {it + 1}.groupBy {lp.mp(it)} // Group numbers by their max prime factor
        val ans = IntArray(n)
        for ((_, v) in a) {
            // For each group, assign elements cyclically to minimize fixed points
            for (i in v.indices) {
                ans[v[i] - 1] = v[(i + 1) % v.size]
            }
        }
        println(ans.joinToString(" "))
    }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/2123_F_Minimize_Fixed_Points.py
# Problem: CF 2123 F - Minimize Fixed Points
# https://codeforces.com/contest/2123/problem/F

"""
F. Minimize Fixed Points
Algorithm: Sieve of Eratosthenes + Permutation Construction

Purpose:
This code finds a "good" permutation of length n (where gcd(p[i], i) > 1 for all 2 <= i <= n)
with the minimum number of fixed points. The approach uses a precomputed sieve to determine
the smallest prime factor of each number, then constructs the permutation using a specific
swapping strategy that minimizes fixed points.

Time Complexity: O(n log log n) for sieve + O(n) for permutation construction = O(n log log n)
Space Complexity: O(n) for storing the sieve and intermediate arrays

The algorithm leverages the fact that for a permutation to be good, each element at position i
must satisfy gcd(p[i], i) > 1. This implies p[i] and i share a common prime factor.
We use smallest prime factor (SPF) decomposition to efficiently group numbers that share factors.

"""

is_prime = [k for k in range(200001)]  # Initialize array with indices


def prime():
    is_prime[0] = is_prime[1] = 1  # Mark 0 and 1 as not prime
    for i in range(2, 200001):
        if is_prime[i] == i:  # If i is prime
            for j in range(i, 200001, i):  # Mark multiples of i with SPF = i
                is_prime[j] = i


prime()  # Precompute smallest prime factors

t = int(input())
for l in range(t):
    n = int(input())
    ans = [0] * (n + 1)
    jump = [k for k in range(n + 1)]  # Jump array to track mappings
    ans[1] = 1  # Base case: p[1] = 1
    for i in range(n + 1):
        jump[i] = jump[is_prime[i]]  # Update jump based on smallest prime factor
        jump[is_prime[i]] = i       # Reverse mapping to track groupings
    for i in range(n + 1):
        ans[jump[i]] = i            # Final permutation assignment
    for i in range(1, n + 1):
        print(ans[i], end=" ")      # Output the result
    print()  # New line after each test case


# https://github.com/VaHiX/CodeForces/